### ISSUE TITLE
Shrinkwrap is forcing insecure dependencies

### ISSUE THREAD

**Issue by isaacs** [netlify/cli#6731]
### Describe the bug

Because this module ships with a shrinkwrap, it is impossible to avoid using the insecure `braces@3.0.2`, even using overrides.

Please either keep all dependencies up to date with security advisories, or remove the shrinkwrap from the deployment so that we can get security updates.

### Steps to reproduce

- npm install netlify-cli
- npm audit

Observe that `braces@3.0.2` is present in the tree, and cannot be overridden, due to the npm-shrinkwrap.json

### Configuration

Not relevant.

### Environment

  System:
    OS: macOS 14.5
    CPU: (16) arm64 Apple M3 Max
    Memory: 14.46 GB / 128.00 GB
    Shell: 5.2.0 - /usr/local/bin/bash
  Binaries:
    Node: 20.13.1 - /usr/local/bin/node
    Yarn: 1.22.18 - /usr/local/bin/yarn
    npm: 10.7.0 - /usr/local/bin/npm
    pnpm: 9.2.0 - /usr/local/bin/pnpm
    bun: 1.1.6 - ~/bin/bun
  npmPackages:
    netlify-cli: ^17.29.0 => 17.29.0

**Comment by merlyn-at-netlify** [comment]
Thanks, @isaacs. For context the impact is not very severe given the context that Netlify CLI runs in:

> Affected versions of this package are vulnerable to Uncontrolled resource consumption due improper limitation of the number of characters it can handle, through the parse function. An attacker can cause the application to allocate excessive memory and potentially crash by sending imbalanced braces as input.

From: https://security.snyk.io/package/npm/braces/3.0.2

**Comment by isaacs** [comment]
Yeah, it's more annoying than actually hazardous, for sure. But such is life in a world that incentivizes getting your name on a CVE more than it does the subtle nuance of actual security.

**Comment by G-Rath** [comment]
Related:
  - https://github.com/netlify/cli/issues/6508
  - https://github.com/netlify/cli/issues/6739
  - https://github.com/netlify/cli/pull/6704
  - https://github.com/npm/cli/issues/7356

I'd really love to see the shrinkwrapping go, as it makes our live a lot harder and would let me make a very solid case for moving us off Vercel as they pin their direct dependencies while also [remaining very behind the ball](https://github.com/vercel/vercel/issues/11543)

**Comment by G-Rath** [comment]
there's now a new micromatch vulnerability that we cannot update due to the shrinkwrap: GHSA-952p-6rrq-rcjv

cc @sarahetter

**Comment by electrovir** [comment]
This package has another vulnerability that I am unable to workaround because of `npm-shrinkwrap.json`: https://github.com/advisories/GHSA-52f5-9888-hmc6

Please stop publishing this package with `npm-shrinkwrap.json`: I simply cannot use it.

**Comment by G-Rath** [comment]
Have opened #7741 about another downstream vulnerability that is blocked due to the shrinkwrap

**Comment by G-Rath** [comment]
... and now the use of shrinkwrap is breaking `npm ci` because of platform constraints: https://github.com/netlify/cli/issues/7769

**Comment by taylorreece** [comment]
Yeah... this shrinkwrap thing is getting pretty annoying. I'm awaiting a release that includes https://github.com/netlify/cli/pull/7758 which was patched last week, but the currently available version is still shrinkwrapped to the old version of `tar`. With the shrinkwrap file, there's no way to `npm audit --fix` the vulnerability. 

Netlify team - any plans to remove the shrinkwrap file? NPM docs recommend against including it https://docs.npmjs.com/cli/v11/configuring-npm/npm-shrinkwrap-json

> It's strongly discouraged for library authors to publish this file, since that would prevent end users from having control over transitive dependency updates.

**Comment by serhalp** [comment]
@taylorreece but note the previous sentence:

> The recommended use-case for npm-shrinkwrap.json is applications deployed through the publishing process on the registry: for example, daemons and command-line tools intended as global installs or devDependencies.

This is exactly the use case we have with netlify-cli.

**Comment by serhalp** [comment]
> I'm awaiting a release that includes https://github.com/netlify/cli/pull/7758 which was patched last week

🤔 No release PR was created because Dependabot marked that as `chore:`. Let me look into that...

**Comment by isaacs** [comment]
By the way, I actually have come to feel like this is now not "merely annoying", it is actually a security issue.

If you use the netlify-cli as part of your deployment, you must have audit failures.

So you come here, find out that it's no biggie, and start ignoring it.

Now you ignore audit failures, and every time audit fails, you go "Oh, right, it's no big deal".

Which it is, until it isn't.

**Comment by isaacs** [comment]
This is an ever-growing problem. All my audit warnings are because I'm using netlify. It's a bad look! Just remove the shrinkwrap, there's no need for it.

I have fixed several _really bad security vulnerabilities_ recently in node-tar, and you're just not allowing your users to benefit from them.

**Comment by G-Rath** [comment]
Not to mention you're also punishing npm users for choosing to use npm, since other package managers ignore the shrink-wrap (and by extension proving the shrink wrap can't be that important).

@isaacs I'm curious on how viable you think it would be to try and add a new `.npmrc` option to npm for having shrinkwraps ignored

**Comment by isaacs** [comment]
@G-Rath I think shrinkwraps were a mistake, and npm should drop support for them in the next major.

However, I no longer work on npm, so I have no more power to effect this than anyone else.

**Comment by G-Rath** [comment]
@isaacs thanks yeah I was asking if you would think it would be tons of work to change or really straightforward, based on what you can remember of how shrinkwrapping is implemented (I assume its not load bearing, but you never know...)

I've created https://github.com/npm/cli/issues/8938 to give us a place to discuss this in the more general npm context

**Comment by isaacs** [comment]
Oh yeah, it'd be crazy easy to just stop doing shrink wraps. The hard part is just accepting the breaking change, but that's social, not technical.
